common/matrix/batch_dense_kernels.hpp.inc

```diff
@@ -166,15 +168,44 @@ __device__ __forceinline__ void compute_norm2(
     const gko::batch_dense::BatchEntry<remove_complex<ValueType>> &result,
     const uint32 &converged)
 {
-    for (int c = 0; c < x.num_rhs; c++) {
-        const uint32 conv = converged & (1 << c);
+    using real_type = typename gko::remove_complex<ValueType>;
+    constexpr auto tile_size = config::warp_size;
+
+    auto thread_block = group::this_thread_block();
+
+    auto subwarp_grp = group::tiled_partition<tile_size>(thread_block);
+
+    const auto subwarp_grp_id = static_cast<int>(threadIdx.x / tile_size);
+
+    const int num_subwarp_grps_per_block = ceildiv(blockDim.x, tile_size);
+
+    for (int rhs_index = subwarp_grp_id; rhs_index < x.num_rhs;
+         rhs_index += num_subwarp_grps_per_block) {
+        const uint32 conv = converged & (1 << rhs_index);
+
         if (conv) {
             continue;
         }
-        one_norm2(x, c, result.values);
+
+        real_type val = zero<real_type>();
+
+        for (int r = subwarp_grp.thread_rank(); r < x.num_rows;
+             r += subwarp_grp.size()) {
+            val += squared_norm(x.values[r * x.stride + rhs_index]);
+        }
+
+        // subwarp_grp level reduction
+        for (int j = subwarp_grp.size() / 2; j > 0; j /= 2) {
+            val += subwarp_grp.shfl_down(val, j);
+        }
+
+        if (subwarp_grp.thread_rank() == 0) {
+            result.values[rhs_index] = sqrt(val);
+        }
```

Comment:
Please put this in a separate `__device__ __forceinline__` function. That was the purpose of the `one_norm2` function earlier.

Reply:
Done!